pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -175,67 +166,127 @@ public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanc
     if (currentAssignment.equals(targetAssignment)) {
       LOGGER.info("Table: {} is already balanced", tableNameWithType);
       if (reassignInstances) {
-        return new RebalanceResult(RebalanceResult.Status.DONE, "Instance reassigned, table is already balanced",
-            instancePartitionsMap, targetAssignment);
+        if (dryRun) {
+          return new RebalanceResult(RebalanceResult.Status.DONE,
+              "Instance reassigned in dry-run mode, table is already balanced", instancePartitionsMap,
+              targetAssignment);
+        } else {
+          return new RebalanceResult(RebalanceResult.Status.DONE, "Instance reassigned, table is already balanced",
+              instancePartitionsMap, targetAssignment);
+        }
       } else {
         return new RebalanceResult(RebalanceResult.Status.NO_OP, "Table is already balanced", instancePartitionsMap,
             targetAssignment);
       }
     }
 
     if (dryRun) {
-      LOGGER.info("Rebalance table: {} in dry-run mode, returning the target assignment", tableNameWithType);
+      LOGGER.info("Rebalancing table: {} in dry-run mode, returning the target assignment", tableNameWithType);
       return new RebalanceResult(RebalanceResult.Status.DONE, "Dry-run mode", instancePartitionsMap, targetAssignment);
     }
 
-    int minAvailableReplicas;
-    if (rebalanceConfig.getBoolean(RebalanceConfigConstants.DOWNTIME, RebalanceConfigConstants.DEFAULT_DOWNTIME)) {
-      minAvailableReplicas = 0;
+    if (!downtime && !currentIdealState.isEnabled()) {
+      LOGGER.warn("Table: {} is disabled, rebalancing it with downtime", tableNameWithType);
+      downtime = true;
+    }
+
+    if (downtime) {
       LOGGER.info("Rebalancing table: {} with downtime", tableNameWithType);
-    } else {
-      minAvailableReplicas = rebalanceConfig.getInt(RebalanceConfigConstants.MIN_REPLICAS_TO_KEEP_UP_FOR_NO_DOWNTIME,
-          RebalanceConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEP_UP_FOR_NO_DOWNTIME);
-      int numCurrentReplicas = currentAssignment.values().iterator().next().size();
-      int numTargetReplicas = targetAssignment.values().iterator().next().size();
-      // Use the smaller one to determine the min available replicas
-      int numReplicas = Math.min(numCurrentReplicas, numTargetReplicas);
-      if (minAvailableReplicas > 0) {
-        if (minAvailableReplicas >= numReplicas) {
-          LOGGER.warn(
-              "Illegal config for min available replicas: {} for table: {}, must be less than number of replicas (current: {}, target: {})",
-              minAvailableReplicas, tableNameWithType, numCurrentReplicas, numTargetReplicas);
-          return new RebalanceResult(RebalanceResult.Status.FAILED, "Illegal min available replicas config",
+
+      while (true) {
```

Comment:
Is is useful to extract this part into a method (provided it has reasonable number of arguments. Similarly the else part.

Reply:
Too many arguments need to be passed, and it is called only once, so prefer keeping it inline for now